### PR TITLE
Bug 1318326 - Fix `get_pings_properties` for keyedHistograms

### PR DESCRIPTION
**Pull request by acmiyaguchi**
[Bug 1318326 - |get_pings_properties| does not return all the available keyedHistograms](https://bugzilla.mozilla.org/show_bug.cgi?id=1318326)

This PR introduces failing tests that demonstrate the current behavior of `get_pings_properties` when retrieving keyedHistograms in pings that contain child processes. This is solved by defaulting to empty dictionaries to avoid `KeyError` exceptions.

---
This change is [Reviewable](https://reviewable.io/reviews/mozilla/python_moztelemetry/150)

### CODE REVIEW

**Comment by acmiyaguchi** [comment]
The updated PR fixes test failures due to bug 1363934. This also adds one more edge case that wasn't caught in the initial PR. 

The updated notebook that verifies the results can be found [here](https://gist.github.com/acmiyaguchi/0f222374f9f622b4fc3415987811ffb5).

**Comment by Dexterp37** [comment]
This looks good to me, but @fbertsch might be a more appropriate reviewer as I haven't touched this code in ages.

File: moztelemetry/spark.py [hunk 1]
```diff
@@ -182,16 +182,20 @@ def _get_ping_properties(ping, paths, only_median, with_processes,
                                            .get("content", {}) \
                                            .get("keyedHistograms", None)
                         if content_kh is not None:
-                            kh_keys.update(content_kh[path[1]].keys())
+                            kh_keys.update(content_kh.get(path[1], {}).keys())
                         else:
                             for payload in cursor.get("childPayloads", []):
-                                kh_keys.update(payload["keyedHistograms"][path[1]].keys())
+                                kh_keys.update(
```
Comment (fbertsch) [inline moztelemetry/spark.py]: Please replace all uses of update with `kh_keys |= someDict.viewkeys()`; it makes it more apparent this is a set and not a dict.

File: moztelemetry/spark.py [hunk 2]
```diff
@@ -182,16 +182,20 @@ def _get_ping_properties(ping, paths, only_median, with_processes,
                                            .get("content", {}) \
```
Comment (fbertsch) [inline moztelemetry/spark.py]: Mauro and I were discussing this and think that using a safe dict as the cursor might be a better (and more readable) solution. See [What is the best way to implement nested dictionaries in Python?](http://stackoverflow.com/questions/635483/what-is-the-best-way-to-implement-nested-dictionaries/19829714#19829714)

Then we can replace all `.get(elem)` with just `[elem]`

File: moztelemetry/spark.py [hunk 3]
```diff
@@ -182,16 +182,20 @@ def _get_ping_properties(ping, paths, only_median, with_processes,
...
+                                        .get(path[1], {})
+                                        .keys())
 
                         gpu_kh = cursor.get("processes", {}) \
                                        .get("gpu", {}) \
                                        .get("keyedHistograms", None)
                         if gpu_kh is not None:
-                            kh_keys.update(gpu_kh[path[1]].keys())
+                            kh_keys.update(gpu_kh.get(path[1], {}).keys())
                 except:
```
Comment (acmiyaguchi) [inline moztelemetry/spark.py]: I'm all for dropping this exception, it seems to be just catching KeyErrors.

**Comment by acmiyaguchi** [comment]
@fbertsch This is ready for another look. I've implemented the safe dictionary and fixed any places where it could have changed the logic. I looked over the Histogram implementation, but I don't think the new KeyError-less behavior affects anything there.

File: moztelemetry/spark.py [hunk 4]
```diff
@@ -162,39 +180,34 @@ def _get_ping_properties(ping, paths, only_median, with_processes,
...
-                            kh_keys.update(gpu_kh[path[1]].keys())
-                except:
-                    result[property_name] = None
-                    continue
+                # These are returned as a subdict mapped from the
+                # property_name.
+                kh_keys = cursor["keyedHistograms"][path[1]].viewkeys()
+                if isinstance(cursor, dict):
+                    content_kh = cursor["processes"]["content"]["keyedHistograms"]
+                    if content_kh:
```
Comment (fbertsch) [inline moztelemetry/spark.py]: We don't need these ifs anymore, do we? The safe cursor will return an empty `dict`, where `content_kh[path[1]].viewkeys()` will return an empty set. So you could reduce this to:

`kh_keys |= cursor["processes"]["content"]["keyedHistograms"][path[1]].viewkeys()`

Same with `gpu_kh`.

File: moztelemetry/spark.py [hunk 5]
```diff
@@ -162,39 +180,34 @@ def _get_ping_properties(ping, paths, only_median, with_processes,
...
-                        if gpu_kh is not None:
-                            kh_keys.update(gpu_kh[path[1]].keys())
-                except:
-                    result[property_name] = None
-                    continue
+                # These are returned as a subdict mapped from the
+                # property_name.
+                kh_keys = cursor["keyedHistograms"][path[1]].viewkeys()
+                if isinstance(cursor, dict):
+                    content_kh = cursor["processes"]["content"]["keyedHistograms"]
```
Comment (fbertsch) [inline moztelemetry/spark.py]: It occurs to me that this will fail if `cursor["processes"]`, `cursor["processes"]["content"]`, etc. are present but not `dict`s. Do you know if that schema is mandated upstream?

Comment (acmiyaguchi) [inline moztelemetry/spark.py]: This should be mandated by the ingestion schema's. Here are the relevant blobs from the main ping schema that defines `processes` and `content` as objects.

https://github.com/mozilla-services/mozilla-pipeline-schemas/blob/master/schemas/telemetry/main/main.4.schema.json#L556-L567
https://github.com/mozilla-services/mozilla-pipeline-schemas/blob/master/schemas/telemetry/main/main.4.schema.json#L875-L876

File: moztelemetry/spark.py (hunk 4 above)
Comment (acmiyaguchi) [inline moztelemetry/spark.py]: That's true, I'll make comment of why this was structured this way.

**Comment by acmiyaguchi** [comment]
@fbertsch this is ready for another look
